BTrees/_base.py

```diff
@@ -815,9 +815,7 @@ def __delitem__(self, key):
         self._del(self._to_key(key))
 
     def clear(self):
-        if self._data:
-            # In the case of __init__, this was already set by __new__
-            self._data = []
+        self._data = []
```

Comment:
> I see, thanks. So it is not clear-specific but just a general things about variables in slots. My approach, once again, would be to fix those cases one-by-one with, as you suggested, bumping persistent version for the slots ghostification patch.

Yes, it is `clear` specific. `clear` is (now) being called when the object's state is undefined. The state that's was supposed to be an invariant, established by `__new__` and maintained by `__setstate__`. That invariant is now broken, and calling a method when the invariants of the class do not hold is a sure way to have trouble (not to mention just being rude). Since the invariant is now broken, `__setstate__` should not call into code that *may or may not* rely on that invariant---it has no way of knowing.

It is far better to avoid introducing trouble at the chokepoint, where it can be fixed all at once---the superclass--than expect everyone else to go clean up when their code breaks.

> It would be a pity not to be able to rely on ghostification semantics to completely release objects.

First, we've never been able to rely on that. persistent has never done that. This change obviously breaks a number of cases. We've located two just by looking at a handful of (very widely used) projects. 

It starting to seem to me that changing this and breaking unknown downstream projects is far more of a risk than simply using `__dict__` for objects where allowing `__slots__` to remain is a problem. 

Second, it's *still* not a guarantee. When the object is implemented in C, if it doesn't specifically do something with its struct members, they aren't released. This is the only reason that zope.container's Proxy still works in the C implementation after this change---the struct member that is `_wrapped` is left sitting untouched by `_p_deactivate`.